### ISSUE TITLE
Support glob patterns and factors in 'depends' option

### ISSUE THREAD

**Issue by ziima** [tox-dev/tox#1152]
Addition of parallel runs and dependencies is awesome, but we quite miss any support for factors or glob patterns in the `depends` option.

In case we have several environments
```
[tox]
envlist = clean-coverage,py27,py35,py36,py37,compute-coverage
```
we have to set dependencies for `compute-coverage` individually
```
[testenv:compute-coverage]
depends = py27,py35,py36,py37
```
which is error prone and not quite comfortable.

I suggest to allow glob patterns
```
[testenv:compute-coverage]
depends = py*
```
and factors similarly to other parts of configuration.

I don't quite feel I would be able to provide a PR.

**Comment by gaborbernat** [comment]
Factors are supported (or anything that works within envlist that is). Patterns AFAIK not so much. In principle, I'm not against this. Could be non-trivial to implement it for the general case though.

**Comment by tpazderka** [comment]
Handling of factors is probably not working, see MWE below.

```
[tox]
minversion = 3.7.0
envlist =
    py27-test
    py37-test
    compute-coverage

[testenv]
skip_install = True
whitelist_externals = 
    sleep
    echo
commands =
    sleep 5
    echo "Test ..."

[testenv:compute-coverage]
depends = test
commands =
    echo "Compute ..."
```
and the output:
```
tox -p all
✔ OK compute-coverage in 0.611 second
✔ OK py37-test in 5.526 seconds
✔ OK py27-test in 5.552 seconds
_____________________________________________________________________________________________________ summary _____________________________________________________________________________________________________
  py27-test: commands succeeded
  py37-test: commands succeeded
  compute-coverage: commands succeeded
  congratulations :)
```

**Comment by gaborbernat** [comment]
Ah, like that will not work. That actually is not the factor syntax, that's the mark syntax. Here's what I am referring to when I see factor support:

```
[tox]
envlist={py27,py37}-{test,notest}-{a, b}, other

[testenv:compute-coverage]
depends = {py27,py37}-{test,notest}-{a, b}
```

What you're implying there resembles more along the lines of environment markers:
```
[tox]
envlist={py27,py37}-{test,notest}-{a, b}, other

[testenv]
deps = test: a
       notest: b
```

That way on its own cannot work because what if there is also a test env. Does test refer to that one env, or all envs having the test factor? At best I could see something like ``{test}`` which would indicate it's a factor. This would be though expending the syntax for factor specification, that would only exist for depends.

**Comment by ziima** [comment]
According to tox documentation: "Parts of an environment name delimited by hyphens are called factors ...", we meant factors (I hope so). But the way we want to use them is different from all other keywords. Thanks for pointing it out. That really would not be easy.

The main problem we want to resolve is the need to copy sublists of environments. When new environment is added, but forgotten in `depends`, spurious race conditions may occur. That's why we want to use glob patterns, e.g. `py*`, or factors (sort of), e.g. `quality` for `quality-*`.

**Comment by gaborbernat** [comment]
Agreed is a gap, contributions are welcome, otherwise, hopefully, we'll get to this at some point.

**Comment by rpkilby** [comment]
An implementation here might also be able to solve #318 at the same time.

**Comment by gaborbernat** [comment]
Currently, we do support factors and label based selection. Glob pattern could be added, perhaps would be under the `-g` flag :+1:

**Comment by gaborbernat** [comment]
This is now available in [version 4.36.0](https://github.com/tox-dev/tox/releases/tag/4.36.0) 🎉
